### ISSUE TITLE
Discrete Integral Transform (DIT) for rapid opacity computation for many lines

### ISSUE THREAD

**Issue by HajimeKawahara** [HajimeKawahara/exojax#88]
I'd like to include [DIT](https://www.sciencedirect.com/science/article/abs/pii/S0022407320310049) into exojax. Starting from a separated repo, [addit](https://github.com/HajimeKawahara/addit), then I plan to import some modules into exojax.

**Comment by HajimeKawahara** [comment]
Started to import functionality from [addit](https://github.com/HajimeKawahara/addit) in [this branch](https://github.com/HajimeKawahara/exojax/tree/features/dit).

**Comment by HajimeKawahara** [comment]
Added an example using HITRAN/CO (Cross Section using Discrete Integral Transform.ipynb). 
![image](https://user-images.githubusercontent.com/15956904/124553880-46c5be80-de70-11eb-98c8-4b4aefb1b1ed.png)

**Comment by HajimeKawahara** [comment]
Also, added the hybrid (DIT+[LPF](http://secondearths.sakura.ne.jp/exojax/exojax/exojax.spec.html#module-exojax.spec.lpf)) case (Hybrid Cross Section using lpf and dit.ipynb)
![image](https://user-images.githubusercontent.com/15956904/124561116-7c6ea580-de78-11eb-9673-49b6fa974dce.png)

![image](https://user-images.githubusercontent.com/15956904/124561145-855f7700-de78-11eb-9af0-be55c1c2dcd1.png)

The hybrid model provides a smaller difference between the direct [lpf](http://secondearths.sakura.ne.jp/exojax/exojax/exojax.spec.html#module-exojax.spec.lpf).

**Comment by HajimeKawahara** [comment]
Forward modeling w/ DIT+LPF has been tested. (Forward modeling using the Hybrid Cross Section.ipynb)

DIT grid (only 4 grids for each layer because the range is narrow)
![image](https://user-images.githubusercontent.com/15956904/124694990-46cfc800-df1d-11eb-8aaa-cf4d652fb010.png)

Cross section matrix
![image](https://user-images.githubusercontent.com/15956904/124695002-4fc09980-df1d-11eb-918d-52b26a97ea1b.png)

Spectrum
![image](https://user-images.githubusercontent.com/15956904/124695020-58b16b00-df1d-11eb-9d85-6e1635155312.png)

Just used interp to convert ESLIN (linear nu grid) to ESLOG (log nu grid) to apply the response.

**Comment by HajimeKawahara** [comment]
Added the case for methane (Forward modeling using the Hybrid Cross Section for methane)
![image](https://user-images.githubusercontent.com/15956904/124699379-a7fb9980-df25-11eb-99d2-2af70110f5c1.png)

N~10000,

**Comment by HajimeKawahara** [comment]
10% deviation is probably too large. Need to check it.

**Comment by bmorris3** [comment]
Hi @HajimeKawahara, I'm experimenting a bit with your changes in #91, and I'm hitting an error when I try to run [the hybrid cross-section tutorial](https://github.com/HajimeKawahara/exojax/blob/develop/examples/tutorial/Hybrid%20Cross%20Section%20using%20lpf%20and%20dit.ipynb)
```python
---------------------------------------------------------------------------
FilteredStackTrace                        Traceback (most recent call last)
<ipython-input-8-6e67e096312a> in <module>
----> 1 xsdit=dit.xsvector(mdbCO.nu_lines,sigmaD,gammaL,Sij,nus,sigmaD_grid,gammaL_grid)

//anaconda3/envs/pymc/lib/python3.8/site-packages/jaxlib/pocketfft.py in pocketfft(c, a, fft_type, fft_lengths)
     92     builder.PrependUint64(d)
---> 93   pocketfft_shape = builder.EndVector(n)
     94 

FilteredStackTrace: TypeError: EndVector() takes 1 positional argument but 2 were given
```
Any tips on what might be wrong here? Thanks!

**Comment by bmorris3** [comment]
Update: updating and reinstalling jax/jaxlib fixed the above problem!

**Comment by HajimeKawahara** [comment]
With the help of @erwanp and @dcmvdbekerom, I figured out the reason of the difference between DIT and LPF.

1. Coarse wavenumber grid
2. Coarse sigmaD and gammaL grids in DIT. The default should be 0.1 for log10 sigmaD and 0.2 for log10 gammaL.
3. truncation error of wavenumber. Recall our raw wavenumber is like 2000.0001 cm-1, which is close to the machine accuracy for float32 (used in JAX/GPU).

I have two solutions for 3. One is use numpy flaot 64 for wavenumber grid in DIT assuming we do not change it (tagged by "precomputed"). Another is just to remove some offset from wavenumber units, denoted by "3D" (we use 3D grids in JAX/float32 here).   This is the comparisons for these.  The last one (large truncation error means I used wavenumber directly for "3D".

Both the above solutions the difference of below 1%.

![fine_grid](https://user-images.githubusercontent.com/15956904/124915315-5c7fe300-e02c-11eb-8692-86e866355f94.png)

**Comment by HajimeKawahara** [comment]
Another issue was GPU memory limitation for the radiative transfer. This was due to a batch computation of FFT in JAX for all the atmospheric layers. Using lax.scan I changed this computation for layer by layer. This enabled us much wider wavenumber grid even for N=10,000-100,000 w/ my laptop. Here is an example of the emission spectrum w/ Exomol methane (Forward modeling using the DIT Cross Section for methane.ipynb)
![ch4](https://user-images.githubusercontent.com/15956904/125061645-80a0fa00-e0e8-11eb-964c-eb5738e27951.png)

I also noticed in the cross section matrix (N layer x N wavenumber), there are some small negative values close to 0 probably due to errors from FFT. These negative values are small, but affect radiative transfer and results in NaN in the final spectrum. So,  a temporal solution is just to fill 0 for such positions (see  Forward modeling using the DIT Cross Section for methane.ipynb for example).

Because the DIT now provides quite accurate results, I almost eliminate the hybrid model (DIT+LPF) for real use. 
For the small number of lines, we would use LPF, for larger one we can use DIT.

**Comment by HajimeKawahara** [comment]
Unknown small wavy shape found in an emission spectrum in DIT

|xsmode | | | |
| ---- | ---- | ---- | ---- |
|**DIT**|![autodit0](https://user-images.githubusercontent.com/15956904/125151717-07a2b080-e183-11eb-813f-1671162e6f17.png)|![autodit1](https://user-images.githubusercontent.com/15956904/125151718-0bcece00-e183-11eb-9425-990dd60ee691.png)|![autodit2](https://user-images.githubusercontent.com/15956904/125151720-0d989180-e183-11eb-84b9-69df82d93f87.png)|
|**LPF**|![autolpf0](https://user-images.githubusercontent.com/15956904/125151702-f9ed2b00-e182-11eb-985e-c718754e3c6b.png)|![autolpf1](https://user-images.githubusercontent.com/15956904/125151704-fbb6ee80-e182-11eb-8ac5-73073b996328.png)|![autolpf2](https://user-images.githubusercontent.com/15956904/125151808-cced4800-e183-11eb-9c67-dc4470ad5143.png)|

**Comment by dcmvdbekerom** [comment]
We had a very similar problem in RADIS, see details in [this discussion](https://github.com/radis/radis/issues/186#issuecomment-764465580)

In RADIS we solved the issue by checking whether the shoulders in FT space go below a certain threshold, and if not, add another fold until the shoulders are sufficiently small (see snippet below). 
```
def calc_gV_FT(x,wG,wL,dv,folding_thresh):

    gV_FT = lambda x,wG,wL: np.exp(-((np.pi*x*wG)**2/(4*np.log(2)) + np.pi*x*wL))
    x_fold = (x,x[::-1])
    
    result = gV_FT(x,wG,wL)
    n = 1
    while gV_FT(n/(2*dv),wG,wL) >= folding_thresh:
        result += gV_FT(n/(2*dv) + x_fold[n&1],wG,wL)
        n += 1

    return result
```
Note that while this increases the number of lineshapes that need to be calculated, it doesn't increase the number of FT's, so you will barely notice any performance hits for having to calculate a couple lineshapes instead of just one.

**Comment by HajimeKawahara** [comment]
@dcmvdbekerom Thank you for the clear explanation! That is the perfect solution for this issue. The wavy feature completely disappeared after the folding: 

|xsmode | | | |
| ---- | ---- | ---- | ---- |
|**DIT w/ folding**|![autoDIT_fold0](https://user-images.githubusercontent.com/15956904/125183210-9846c280-e24f-11eb-97bb-da7786cc38bc.png)|![auto_DITfold](https://user-images.githubusercontent.com/15956904/125183078-61bc7800-e24e-11eb-91b2-91bb094b1be4.png)|![auto_DITfold2](https://user-images.githubusercontent.com/15956904/125183214-a1379400-e24f-11eb-87b0-729b378a2997.png)|


Fig. DIT w/ the folding 

I did [some tests in addit](https://github.com/HajimeKawahara/addit/blob/main/snippets/test/DIT%20folded%20voigt.ipynb).
I found negative values around the line center can be removed by the folding, but still remains some negative ones at the position far from the line center, probably from different reasons. But, these looks harmless when just replacing them to zero.

**Comment by dcmvdbekerom** [comment]
Glad to see it solved the wavy features!

I suspect the remaining negative values are due to rounding errors in the FT, in which case manually zeroing them indeed looks like the best option.

There is a small chance that the distribution of the weights by `a` & `(1-a)` breaks down when the intensity of the lineshapes is not the expected value due to folding; to see if this might be the case you could try whether renormalizing the lineshapes helps by adding `result /= result[0]` before returning `result` in the code above.

**Comment by HajimeKawahara** [comment]
Closed along w/ the merge of PR #93.
